master/src/main/scala/org/apache/celeborn/service/deploy/master/network/CelebornRackResolver.scala

```diff
@@ -17,33 +17,80 @@
 
 import com.google.common.base.Strings
 import org.apache.hadoop.conf.Configuration
 import org.apache.hadoop.fs.CommonConfigurationKeysPublic
-import org.apache.hadoop.net.{CachedDNSToSwitchMapping, DNSToSwitchMapping, NetworkTopology, Node, NodeBase, ScriptBasedMapping}
+import org.apache.hadoop.net.{CachedDNSToSwitchMapping, DNSToSwitchMapping, NetworkTopology, Node, NodeBase, ScriptBasedMapping, TableMapping}
 import org.apache.hadoop.util.ReflectionUtils
 
 import org.apache.celeborn.common.CelebornConf
 import org.apache.celeborn.common.internal.Logging
-import org.apache.celeborn.common.util.CelebornHadoopUtils
+import org.apache.celeborn.common.util.{CelebornHadoopUtils, ThreadUtils}
 
 class CelebornRackResolver(celebornConf: CelebornConf) extends Logging {
 
+  private var rackResolveRefreshThreadPool: ScheduledExecutorService = _
+
+  private var rackResolveLastModifiedTime = 0L
+
   private val dnsToSwitchMapping: DNSToSwitchMapping = {
     val conf: Configuration = CelebornHadoopUtils.newConfiguration(celebornConf)
     val dnsToSwitchMappingClass =
       conf.getClass(
         CommonConfigurationKeysPublic.NET_TOPOLOGY_NODE_SWITCH_MAPPING_IMPL_KEY,
         classOf[ScriptBasedMapping],
         classOf[DNSToSwitchMapping])
-    ReflectionUtils.newInstance(dnsToSwitchMappingClass, conf)
+    val mapping = ReflectionUtils.newInstance(dnsToSwitchMappingClass, conf)
       .asInstanceOf[DNSToSwitchMapping] match {
       case c: CachedDNSToSwitchMapping => c
       case o => new CachedDNSToSwitchMapping(o)
     }
+    val refreshInterval = celebornConf.rackResolveRefreshInterval
+    rackResolveRefreshThreadPool =
+      ThreadUtils.newDaemonSingleThreadScheduledExecutor("rack-check-thread")
+    val fileName = mapping match {
+      case _: ScriptBasedMapping =>
+        conf.get(CommonConfigurationKeysPublic.NET_TOPOLOGY_SCRIPT_FILE_NAME_KEY)
+      case _: TableMapping =>
+        conf.get(CommonConfigurationKeysPublic.NET_TOPOLOGY_TABLE_MAPPING_FILE_KEY)
+      case _ => null
+    }
+    if (fileName != null) {
+      var scriptFile = new File(fileName)
+      rackResolveLastModifiedTime = scriptFile.lastModified()
+      rackResolveRefreshThreadPool.schedule(
+        new Runnable {
+          override def run(): Unit = {
+            scriptFile = new File(fileName)
+            if (scriptFile.lastModified() != rackResolveLastModifiedTime) {
+              rackResolveLastModifiedTime = scriptFile.lastModified()
+              mapping.reloadCachedMappings()
+            }
+          }
+        },
+        refreshInterval,
+        TimeUnit.MILLISECONDS)
+    } else {
+      rackResolveRefreshThreadPool.schedule(
+        new Runnable {
+          override def run(): Unit = mapping.reloadCachedMappings()
+        },
+        refreshInterval,
+        TimeUnit.MILLISECONDS)
+    }
+    mapping
+  }
+
+  def stop(): Unit = {
+    if (null != rackResolveRefreshThreadPool) {
+      rackResolveRefreshThreadPool.shutdownNow()
+    }
```

Comment:
A potentially simply solution would be to call `dnsToSwitchMapping.reloadCahedMapping` periodically instead ?

Reply:
I'm not sure if there will be performance degradation when there are enough worker nodes and the refresh interval is short enough.

Reply:
It seems that we should not reload when the rack file has not been modified.

Reply:
In general, I would prefer sticking to exposed public api's.
Having said that, for these two specific implementations, given how widely they are available - makes sense to use it.

I would additionally suggest checking if `fileName` `File` is readable - and if not, fallback to `else` block (the `fileName == null` case: in case it is some subclass which handles extensions - for example, script from hdfs/etc)